### PR TITLE
feat(auth): add group membership field resolver provider

### PR DESCRIPTION
**Pull request by amanda-her**
In this pull request the following topics have been addressed:
- The group membership field resolver has been added to return the groups and native groups of a user. This is required to develop a custom authorizer using the plugin approach.
- Improvement in PolicyEngine class to delegate in the field resolver instead of manually invoking the EntityClient class. 
- Fix method `isGroupMatch` in PolicyEngine class to check that a user belongs to at least one group if policy applies to all groups.

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java:113]: I see the idea here.. nice! 

Do we _need_ to resolve the resource spec at this point, or can we do it later on once we absolutely _know_ that group membership is required. I want to make sure that we avoid making any additional lookups until we absolutely have to

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java:194]: Why would this happen? Have we seen this case occur, where actorUrn is null in the request?

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java:198]: Same question as above here - is there any way we can pass down the ResourceSpec and simple resolve once we need to?

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:50]: Minor: I see we are reusing the concept of "ResourceSpec" here, but I'm not sure we want to. The actor is not usually thought of as a "resource", but is its own special concept. Maybe we have a ResolvedActorSpec and ActorSpec classes of their own?

Comment (jjoyce0510) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:50]: What do you think? 

Otherwise, we may be able to rename ResourceSpec as EntitySpec. 

I want to minimize the amount of confusion we introduce here

Comment (jjoyce0510) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:50]: I mean my preference would be to add a parallel set of concepts: 

Actor and ResolvedActor 

or something similar, to keep the two concepts clearly distinct.

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:275]: Nice!

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:406]: I really like that we are able to remove this from here!! Amazing.

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/GroupMembershipFieldResolverProvider.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/GroupMembershipFieldResolverProvider.java:75]: Nice !! Needs unit test

Comment (jjoyce0510) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/GroupMembershipFieldResolverProvider.java:75]: (We are requiring new self-contained java classes have them)

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java
Comment (amanda-her) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java:113]: So there are two options:
- As it is implemented now, resolving the actor and pass it to the PolicyEngine. It is true that at this point we do not know if group membership is required. However, according to [this line](https://github.com/datahub-project/datahub/blob/master/metadata-auth/auth-api/src/main/java/com/datahub/authorization/FieldResolver.java#L16), no field is resolved until it is explicitly called/obtained.
- The other option implies passing the `_resourceSpecResolver ` to the PolicyEngine class. The PolicyEngine class is initialized in the DatahubAuthorizer constructor while the `_resourceSpecResolver `  in the init function. So if we want to follow this approach, the PolicyEngine class should be initialized in the init method after the resolver. Not sure if this has any implications.

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java
Comment (amanda-her) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java:194]: This logic was moved from the [PolicyEngine class](https://github.com/datahub-project/datahub/blob/master/metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java#L50-L63) to here (due to the decision of passing the resolved spec which need a urn).

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java
Comment (amanda-her) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java:198]: The same answer applies here.

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (amanda-her) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java]: We have decided to rename ResourceSpec to EntitySpec (and all other related classes) to avoid the additional complexity of parallel classes with almost the same behavior.

File: metadata-auth/auth-api/src/main/java/com/datahub/authorization/EntityFieldType.java
Comment (jjoyce0510) [inline metadata-auth/auth-api/src/main/java/com/datahub/authorization/EntityFieldType.java:26]: nice!

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java
Comment (jjoyce0510) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/DataHubAuthorizer.java:209]: This is fine - let's go with this. 

I do think in future we will want to make the auth layer a bit more opinionated and work in the URN types instead of raw strings. That was probably a bad call initially

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:112]: actor? to keep naming consistent i'd think

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:133]: actor?

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:160]: thank u!

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:178]: nice!

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:197]: nice renaming. much cleane!

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:112]: (im okay if we do in a followup)

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:242]: seeing this, i am thinking we should always say "actorSpec" and "resourceSpec"

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:267]: do we need this map?

resolvedActorSpec.getSpec().getEntity() is a string right?

File: metadata-service/auth-impl/src/test/java/com/datahub/authorization/PolicyEngineTest.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/test/java/com/datahub/authorization/PolicyEngineTest.java:85]: nice improvements to the test!

File: metadata-service/auth-impl/src/test/java/com/datahub/authorization/PolicyEngineTest.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/test/java/com/datahub/authorization/PolicyEngineTest.java:1187]: strange that we were fetching this

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/EntityFieldResolverProvider.java
Comment (jjoyce0510) [inline metadata-service/auth-impl/src/main/java/com/datahub/authorization/fieldresolverprovider/EntityFieldResolverProvider.java:16]: nice renames!

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (jjoyce0510) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:267]: missed this toString map. nvm

**Review by jjoyce0510 — APPROVED** [review]
Once the conflicts are resolved, let's merge. 

Nice work, @amanda-her !

Really appreciate all the hard work on this.

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (Khurzak) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:112]: Our understanding is that resolvedEntitySpec is preferred as it is consistent with other parameters. However, in this case the parameter must be an actor (the corpuser that is doing the request) and we believe that the current name narrows the purpose of the parameter and helps to understand possible values as not any resolved entity spec can be received.
We hope you agree to keep the current name. In case you still think the rename is a better option we will apply the change on the next PR to avoid blocking this one 🙂

File: metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java
Comment (Khurzak) [reply metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java:133]: Same answer as for the previous comment.

**Comment by Khurzak** [comment]
Hello @jjoyce0510 . We have solved the merge conflicts and think the branch is ready to be merged.

cc @sgomezvillamor

**Comment by jjoyce0510** [comment]
Merging through the flakes on the final smoke test. Unrelated to these changes.

**Pull request merged**
